patches/tModLoader/Terraria/GameContent/UI/Elements/UIAchievementListItem.cs.patch

```diff
@@ -0,0 +1,123 @@
+--- src/TerrariaNetCore/Terraria/GameContent/UI/Elements/UIAchievementListItem.cs
++++ src/tModLoader/Terraria/GameContent/UI/Elements/UIAchievementListItem.cs
+@@ -1,9 +_,11 @@
+ using System;
++using log4net.Repository.Hierarchy;
+ using Microsoft.Xna.Framework;
+ using Microsoft.Xna.Framework.Graphics;
+ using ReLogic.Content;
+ using Terraria.Achievements;
+ using Terraria.Localization;
++using Terraria.ModLoader;
+ using Terraria.UI;
+ using Terraria.UI.Chat;
+ 
+@@ -19,6 +_,7 @@
+ 	private const int _iconsPerRow = 8;
+ 	private int _iconIndex;
+ 	private Rectangle _iconFrame;
++	private float _offsetProgressBarX = 0;
+ 	private Rectangle _iconFrameUnlocked;
+ 	private Rectangle _iconFrameLocked;
+ 	private Asset<Texture2D> _innerPanelTopTexture;
+@@ -29,6 +_,13 @@
+ 
+ 	public UIAchievementListItem(Achievement achievement, bool largeForOtherLanguages)
+ 	{
++		ModAchievement modAchievement = achievement.ModAchievement;
++		if (achievement.Modded) {
++			if (modAchievement == null) {
++				return;
++			}
++		}
++
+ 		_large = largeForOtherLanguages;
+ 		BackgroundColor = new Color(26, 40, 89) * 0.8f;
+ 		BorderColor = new Color(13, 20, 44) * 0.8f;
+@@ -46,9 +_,20 @@
+ 		_iconFrameLocked.X += 528;
+ 		_iconFrame = _iconFrameLocked;
+ 		UpdateIconFrame();
++
++
++		if (achievement.Modded) {
++			_achievementIcon =
++				new UIImageFramed(modAchievement.GetTexture(), new Rectangle((_achievement.IsCompleted ? 0 : 65), 0, 64, 64));
++		}
++		else
++		{
+-		_achievementIcon = new UIImageFramed(Main.Assets.Request<Texture2D>("Images/UI/Achievements"), _iconFrame);
++			_achievementIcon = new UIImageFramed(Main.Assets.Request<Texture2D>("Images/UI/Achievements"), _iconFrame);
++		}
++
+ 		_achievementIcon.Left.Set(num2, 0f);
+ 		_achievementIcon.Top.Set(num3, 0f);
++
+ 		Append(_achievementIcon);
+ 		_achievementIconBorders = new UIImage(Main.Assets.Request<Texture2D>("Images/UI/Achievement_Borders"));
+ 		_achievementIconBorders.Left.Set(-4f + num2, 0f);
+@@ -82,10 +_,18 @@
+ 		float num3 = innerDimensions.Width - dimensions.Width + 1f - (float)(num * 2);
+ 		Vector2 baseScale = new Vector2(0.85f);
+ 		Vector2 baseScale2 = new Vector2(0.92f);
++
++		string descriptionText = _achievement.Description.Value;
++		string nameText = _achievement.FriendlyName.Value;
++		if (_achievement.Hidden && _locked) {
++			nameText = "???";
++			descriptionText = "???";
++		}
```

Comment:
Here's a fun trick, you can use shadowing and value tuples to reduce patch size and make it more maintainable.
```cs
		var _achievement = (this._achievement.Description, this._achievement.FriendlyName, this._achievement.Hidden, this._achievement.Category);

		if (_achievement.Hidden && _locked) {
			_achievement.Description = _achievement.FriendlyName = new LocalizedText(null, "???");
		}
```